modules/sr/robot/utils.py

```diff
@@ -7,3 +20,75 @@ def map_to_range(
 ) -> float:
     """Maps a value from within one range of inputs to within a range of outputs."""
     return ((value - old_min) / (old_max - old_min)) * (new_max - new_min) + new_min
+
+
+def get_robot_emitter(robot: Robot, device: str) -> Emitter:
+    try:
+        # webots 2020b fails to assign the output to the expected type
+        emitter: Optional[Emitter] = robot.getDevice(device)
```

Comment:
Regarding how to handle the type errors you'll be getting here -- we need to either just plain [`cast`](https://docs.python.org/3/library/typing.html#typing.cast) the object, or check it first and then assign it (maybe also with a cast).

I would suggest that we go with an explicit check, in part because we could then extract our own helper function which looks something like this:
``` python
form typing import Type, TypeVar
from controller import Device

TDevice = TypeVar('TDevice', bound=Device)

def get_robot_device(robot: Robot, device: str, kind: Type[TDevice]) -> TDevice:
    device: Optional[Device] = robot.getDevice(device)
    if not isinstance(device, kind):
        raise TypeError(...)
    return cast(TDevice, device)
```
The trick here is that we're using `Type[T] -> T`  to allow us to operate generically over all possible `Device` types while preserving the type information the caller has provided.
Usage looks like this:
``` python
emitter: Emitter = get_robot_device(robot, 'emitter', Emitter)
```
I've omitted the fallback interactions above as I'm not sure how they behave. Hopefully there's enough information about the types above to get you going though.

Reply:
Since each fallback function is specific to the type being returned they could be added with an if-else tree based on the kind argument.
This method seems like a better solution since all the current functions perform aproximately the same actions.

Reply:
I have a strange issue where on webots-2020b robot.getDevice throws a TypeError but if I wrap that line in a try/except and put the fallback function in the error handler the fallback function isn't executed and the instance test fails.

Reply:
This works but removing the try except throws a TypeError and merging the except and device is None clauses causes that code to never be run.
```python
def get_robot_device(robot: Robot, name: str, kind: Type[TDevice]) -> TDevice:
    device: Optional[Device] = None
    try:
        device = robot.getDevice(name)
    except TypeError:
        pass
    if device is None:
        if kind is Motor:
            device = robot.getMotor(name)
    if not isinstance(device, kind):
        raise TypeError
    return cast(TDevice, device)
```